File: src/design_paradigms/divide_and_conquer/merge_sort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    # Divide the array into two halves
    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    # Merge the sorted halves
    return merge(left_half, right_half)

def merge(left, right):
    """
    Merge two sorted arrays into one sorted array.

    Parameters:
        left (list): The first sorted list.
        right (list): The second sorted list.

    Returns:
        list: A merged and sorted list.
    """
    sorted_array = []
    i = j = 0

    # Compare elements from both halves and merge them in sorted order
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            sorted_array.append(left[i])
            i += 1
        else:
            sorted_array.append(right[j])
            j += 1

    # Append any remaining elements from the left half
    while i < len(left):
        sorted_array.append(left[i])
        i += 1

    # Append any remaining elements from the right half
    while j < len(right):
        sorted_array.append(right[j])
        j += 1

    return sorted_array

File: src-tauri/src/Algorithms/sorting_algorithms/merge_sort.rs
pub fn merge_sort<T: PartialOrd + Clone>(arr: &mut [T]) {
    if arr.len() <= 1 {
        return;
    }

    let mid = arr.len() / 2;

    // Sort the left and right halves first
    merge_sort(&mut arr[..mid]);
    merge_sort(&mut arr[mid..]);

    // Copy sorted halves for merging
    let left_half = arr[..mid].to_vec();
    let right_half = arr[mid..].to_vec();

    merge(arr, &left_half, &right_half);
}

fn merge<T: PartialOrd + Clone>(arr: &mut [T], left: &[T], right: &[T]) {
    let mut i = 0;
    let mut j = 0;
    let mut k = 0;

    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            arr[k] = left[i].clone();
            i += 1;
        } else {
            arr[k] = right[j].clone();
            j += 1;
        }
        k += 1;
    }

    while i < left.len() {
        arr[k] = left[i].clone();
        i += 1;
        k += 1;
    }

    while j < right.len() {
        arr[k] = right[j].clone();
        j += 1;
        k += 1;
    }
}
